### PR TITLE
Properly flush unique queues on startup

### LINKED ISSUE #23050: Queues not terminating gracefully
### Description

We are running into the issue where a Gitea restart happens during merge conflict checking, PRs remain stuck in conflict state forever. I think there's two distinct issues here, one is that these don't get unstuck on new pushes and perhaps that's best left for another report.

However the reason things get into this state in the first place seems to be a problem in shutdown terminating all the queues immediately instead of respecting the hammer time.

When starting `./gitea web` and then do `killall gitea` , I see the following in the log:
```
2023/02/21 19:43:26 ...eful/manager_unix.go:149:handleSignals() [W] [63f510cd-4] PID 1500309. Received SIGTERM. Shutting down...
2023/02/21 19:43:26 cmd/web.go:271:listen() [I] [63f510cd-48] HTTP Listener: 0.0.0.0:3000 Closed
2023/02/21 19:43:26 ...eue/queue_channel.go:127:func1() [W] ChannelQueue: task-channel Terminated before completed flushing
2023/02/21 19:43:26 .../graceful/manager.go:205:doHammerTime() [W] Setting Hammer condition
2023/02/21 19:43:26 ...eue/queue_channel.go:127:func1() [W] ChannelQueue: notification-service-channel Terminated before completed flushing
2023/02/21 19:43:26 ...eue/queue_channel.go:127:func1() [W] ChannelQueue: push_update-channel Terminated before completed flushing
2023/02/21 19:43:26 ...eful/server_hooks.go:46:doShutdown() [I] [63f510cd-48] PID: 1500309 Listener ([::]:3000) closed.
2023/02/21 19:43:27 .../graceful/manager.go:224:doTerminate() [W] Terminating
2023/02/21 19:43:27 ...er/issues/indexer.go:201:2() [I] PID: 1500309 Issue Indexer closed
2023/02/21 19:43:27 ...eful/manager_unix.go:157:handleSignals() [W] PID: 1500309. Background context for manager closed - context canceled - Shutting down...
2023/02/21 19:43:27 cmd/web.go:183:runWeb() [I] PID: 1500309 Gitea Web Finished
```

This log is from my local test instance with configuration set to defaults as much as possible.

On our production instance that `Terminated before completed flushing` is leading to a lot of different errors as workers get terminated in the middle of what they're doing.

I can provide more detailed logs if needed, but maybe this is easy to redo on any instance.

### Gitea Version

main (43405c3)

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Own build from main.

### Database

None

### PR DESCRIPTION
**Pull request by zeripath**
There have been a number of reports of PRs being blocked whilst being checked which have been difficult to debug. In investigating #23050 I have realised that whilst the Warn there is somewhat of a miscall there was a real bug in the way that the LevelUniqueQueue was being restored on start-up of the PersistableChannelUniqueQueue.

Next there is a conflict in the setting of the internal leveldb queue name - This wasn't being set so it was being overridden by other unique queues.

This PR fixes these bugs and adds a testcase.

Thanks to @brechtvl  for noticing the second issue.

Fix #23050
and others

### CODE REVIEW

File: modules/queue/unique_queue_disk_channel_test.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"github.com/stretchr/testify/assert"
```
Comment (zeripath) [inline modules/queue/unique_queue_disk_channel_test.go:15]:
```suggestion

	"github.com/stretchr/testify/assert"
```

**Comment by brechtvl** [comment]
Even with these changes the steps described in https://github.com/go-gitea/gitea/issues/23050#issuecomment-1440573511 still get PRs stuck in conflict checking.

I noticed that in the `levelqueue.UniqueQueue` implementation, `queue.set.Remove` is done after the callback function has completed successfully. This means calling `Has()` from inside the callback function will  always return true, which is not what `checkAndUpdateStatus` expects?
https://gitea.com/lunny/levelqueue/src/branch/master/uniquequeue.go

**Comment by brechtvl** [comment]
It seems `luq.IsEmpty()` returns true in this new code. I tried to figure out why, and looking at contents of the database with `leveldbutil dump 000231.ldb` gave me this:
https://gist.github.com/brechtvl/b2b4710fe0fbffba8242a40f539adbba

It seems to be missing a prefix for different queue types, which causes different queues to overwrite each others `low` and `high` values. These are used to compute the length of the queue, which is used to determine if it's empty.

**Comment by brechtvl** [comment]
With this additional change it works for me:
```
diff --git a/modules/queue/unique_queue_disk_channel.go b/modules/queue/unique_queue_disk_channel.go
index 7934342..ec1c3c0 100644
--- a/modules/queue/unique_queue_disk_channel.go
+++ b/modules/queue/unique_queue_disk_channel.go
@@ -94,7 +94,8 @@ func NewPersistableChannelUniqueQueue(handle HandlerFunc, cfg, exemplar interfac
                        },
                        Workers: 0,
                },
-               DataDir: config.DataDir,
+               DataDir:   config.DataDir,
+               QueueName: config.Name,
        }

        queue.channelQueue = channelUniqueQueue.(*ChannelUniqueQueue)
```

**Comment by zeripath** [comment]
@brechtvl Oh bugger. 

```
diff --git a/modules/queue/unique_queue_disk_channel.go b/modules/queue/unique_queue_disk_channel.go
index 7934342..ec1c3c0 100644
--- a/modules/queue/unique_queue_disk_channel.go
+++ b/modules/queue/unique_queue_disk_channel.go
@@ -94,7 +94,8 @@ func NewPersistableChannelUniqueQueue(handle HandlerFunc, cfg, exemplar interfac
                        },
                        Workers: 0,
                },
-               DataDir: config.DataDir,
+               DataDir:   config.DataDir,
+               QueueName: config.Name,
        }

        queue.channelQueue = channelUniqueQueue.(*ChannelUniqueQueue)

```

Yup I think you're right there. Let me just find where that went missing...

**Comment by zeripath** [comment]
I would note that https://github.com/go-gitea/gitea/pull/18732 was an attempt to provide helpful commands to interrogate the leveldb queues a long time ago and perhaps we could now re-evaluate the utility of this and get it merged.

**Review by delvh — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

File: modules/queue/unique_queue_disk_channel.go [hunk 2]
```diff
@@ -209,17 +210,29 @@ func (q *PersistableChannelUniqueQueue) Run(atShutdown, atTerminate func(func())
 	atTerminate(q.Terminate)
 	_ = q.channelQueue.AddWorkers(q.channelQueue.workers, 0)
 
-	if luq, ok := q.internal.(*LevelUniqueQueue); ok && luq.ByteFIFOUniqueQueue.byteFIFO.Len(luq.shutdownCtx) != 0 {
+	if luq, ok := q.internal.(*LevelUniqueQueue); ok && !luq.IsEmpty() {
```
Comment (lunny) [inline modules/queue/unique_queue_disk_channel.go:213]: Why we need to add the empty check here?

**Comment by wxiaoguang** [comment]
Caught a flaky test:

https://drone.gitea.io/go-gitea/gitea/70256/2/14

```
--- FAIL: TestPersistableChannelUniqueQueue (0.62s)

    unique_queue_disk_channel_test.go:227: TestPersistableChannelUniqueQueue executedInitiallyA=57, executedInitiallyB=75, executedToEmptyA=44, executedToEmptyB=26

    --- FAIL: TestPersistableChannelUniqueQueue/Empty_Queue:_QueueA#01 (0.17s)

        unique_queue_disk_channel_test.go:210: 

            	Error Trace:	/drone/src/modules/queue/unique_queue_disk_channel_test.go:210

            	Error:      	Not equal: 

            	            	expected: 101

            	            	actual  : 100

            	Test:       	TestPersistableChannelUniqueQueue/Empty_Queue:_QueueA#01

        unique_queue_disk_channel_test.go:211: 

            	Error Trace:	/drone/src/modules/queue/unique_queue_disk_channel_test.go:211

            	Error:      	Not equal: 

            	            	expected: 0

            	            	actual  : 1

            	Test:       	TestPersistableChannelUniqueQueue/Empty_Queue:_QueueA#01

    unique_queue_disk_channel_test.go:256: TestPersistableChannelUniqueQueue executedInitiallyA=21, executedInitiallyB=100, executedToEmptyA=79, executedToEmptyB=1

FAIL
```

**Pull request merged**